### PR TITLE
Add Vulkan QML backend

### PR DESCRIPTION
**Pull request by darksylinc**
# 🎉 New feature

## Order in which PRs must be merged

1. gazebosim/gz-rendering#706
2. gazebosim/gz-gui#357
    - or optionally, merge this one as part of the next one aka PR 467, which includes it
3. This PR.
4. gazebosim/gz-sim#1735

## Summary

> **_Note:_** This PR depends on gazebosim/gz-rendering#706

> **_Note:_** This PR is polluted with changes that are supposed to be merged by PR #357 

This PR adds native Vulkan backend using Qt's QML Vulkan RHI.

I couldn't believe it so I opened it up on RenderDoc and indeed it's using Vulkan:

Screenshot_2022-08-14_19-51-44

This PR breaks ABI slightly, so it's best if it targets post-Garden.

With a few global variables we could workaround it and backport it to Garden though. Note that Garden supports Vulkan via a fallback that is very slow. This PR offers native, full speed integration.

> **_Note:_** Qt 5.14 is the minimum requirement, which means it's not available on Ubuntu 20.04 but rather Ubuntu 22.04. I put macros to use at least 5.15.2 because that's what I tested against.

> **_Note:_** Qt 5 considers Vulkan QML interface to be 'experimental' and I've submitted a [patch to Qt](https://bugreports.qt.io/browse/QTBUG-105158) to improve compliance with the Vulkan standard. Ideally Gazebo should eventually migrate to Qt 6.

## Missing tasks:

- [x] Ensure code compiles without Qt 5.15.2+
- [x] Ensure code compiles when Vulkan isn't available
- [x] Gazebo needs to tell OgreNext to make a layout transition / synchronization for Qt (via BarrierSolver).
- [x] Run Vulkan validations to catch any potential issue
    - Done. Only validation errors show up on shutdown, but that's difficult to debug right now because gz-gui crashes on shutdown due to infamous crash at shutdown (can't find the ticket now) that could be the cause. May also be a side effect of #17 
- [ ] Add vulkan toggle via command line

## Test it

This PR can be tested once gazebosim/gz-sim#1735 (which comes after this PR) is merged, which adds the capability to enable the features in this PR via CLI.

### Testing it in Ubuntu 20.04

It is possible to build with Ubuntu 20.04 against Qt 5.15.2. You need the following:

1. Install Qt 5.15.2 using the [online installer](https://download.qt.io/official_releases/online_installers/)
2. [Compile Gazebo](https://gazebosim.org/docs/garden/install_ubuntu_src) with: `colcon build --cmake-args -DCMAKE_PREFIX_PATH=/opt/Qt/5.15.2/gcc_64 -DQT_QMAKE_EXECUTABLE=/opt/Qt/5.15.2/gcc_64/bin/qmake --merge-install`
3. To run gazebo go to the install folder and do:
```bash
export LD_LIBRARY_PATH=/opt/Qt/5.15.2/gcc_64/lib
. setup.bash
gz sim shapes.sdf
```

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/plugins/minimal_scene/MinimalScene.cc [hunk 1]
```diff
@@ -537,7 +555,88 @@ void GzRenderer::BroadcastKeyPress()
...
+static void fillQtDeviceExtensionsToOgre(
+  rendering::GzVulkanExternalDevice &externalDevice)
+{
+  // We know Qt adds these by looking at
+  //  qt-everywhere-src-5.15.2/qtbase/src/gui/rhi/qrhivulkan.cpp
+  QVector<QString> deviceExtensions;
+  deviceExtensions.append( "VK_KHR_swapchain" );
+
+  QByteArray envExts = qgetenv( "QT_VULKAN_DEVICE_EXTENSIONS" );
+  if( !envExts.isEmpty() )
```
Comment (ahcorde) [inline src/plugins/minimal_scene/MinimalScene.cc:615]:
```suggestion
  if(!envExts.isEmpty())
```

File: src/plugins/minimal_scene/MinimalScene.cc [hunk 2]
```diff
@@ -537,7 +555,88 @@ void GzRenderer::BroadcastKeyPress()
...
+  {
+    QByteArrayList envExtList = envExts.split( ';' );
+    for( const auto &ext : envExtList )
+    {
+      if( !ext.isEmpty() )
```
Comment (ahcorde) [inline src/plugins/minimal_scene/MinimalScene.cc:620]:
```suggestion
    for(const auto &ext : envExtList)
    {
      if(!ext.isEmpty())
```

File: src/plugins/minimal_scene/MinimalSceneRhiVulkan.cc [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+/////////////////////////////////////////////////
+void GzCameraTextureRhiVulkan::Update(rendering::CameraPtr _camera)
+{
+  // It says Metal but it also works for Vulkan in the exact same way
+  _camera->RenderTextureMetalId(&this->dataPtr->textureId);
+}
+
+/////////////////////////////////////////////////
+/////////////////////////////////////////////////
```
Comment (ahcorde) [inline src/plugins/minimal_scene/MinimalSceneRhiVulkan.cc:90]:
```suggestion
```

**Comment by iche033** [comment]
> This PR breaks ABI slightly, so it's best if it targets post-Garden.

I think there is no ABI breakage? Adding a new function to `Helpers.hh` should be ok. The minimal scene plugin headers are not installed.

File: src/plugins/minimal_scene/MinimalSceneRhiVulkan.cc (hunk 3 above)
Comment (darksylinc) [reply src/plugins/minimal_scene/MinimalSceneRhiVulkan.cc:91]: The double comment is on purpose (it's also in MinimalSceneRhiOpenGL.cc.cpp & MinimalSceneRhiMetal.mm) to distinguish the switch from `GzCameraTextureRhiVulkan` class definitions to `RenderThreadRhiVulkan`.

(fun fact: GzCameraTextureRhi* don't seem to actually be used. It seems to be dead code but this theory needs to be tested; I added a TODO in `class GzCameraTextureRhi`).

**Comment by iche033** [comment]
we need to bump gz-rendering dep version to 8.0.0 and update CI configurations. I'll do that in a separate PRs after all changes are forward ported to `main`.

**Comment by mjcarroll** [comment]
> we need to bump gz-rendering dep version to 8.0.0 and update CI configurations. I'll do that in a separate PRs after all changes are forward ported to main.

I thought this could land in garden?

**Comment by iche033** [comment]
> I thought this could land in garden?

https://github.com/gazebosim/gz-rendering/pull/706 may have some ABI breaking changes. We could try to work around it.

**Comment by darksylinc** [comment]
Originally yes. With a few hacks I think/thought it could still land on Garden.

However my original assessment was assuming #357 was already merged (I was surprised when I found out it was not), and AFAIK Garden will be released without it.

I'm not sure if we can avoid ABI breakage given that #357 wasn't included. I'd have to carefully look into it if we want to backport at least the fallback path.

**Comment by iche033** [comment]
I think https://github.com/gazebosim/gz-gui/pull/357 can go in gz-gui6 and it touches only files in the minimal scene GUI plugin. Those headers are not installed so we don't have to worry about API/ABI compatibility.

**Comment by iche033** [comment]
coming back to this, I think https://github.com/gazebosim/gz-rendering/pull/706 is going to be hard to backport to garden. @mjcarroll are you ok with these PRs landing in `main`?

**Comment by mjcarroll** [comment]
Yes, lets land them on main for expedience.

**Comment by azeey** [comment]
needs version bumps

**Comment by darksylinc** [comment]
@azeey I'm trying to rebase this PR to main; but current main has a gazillion build errors.

`gz-gui/main` branch clearly needs to merge with `gz-gui/gz-gui7` branch.

gz-sim/main is in the same state (merging has quite a number of merge conflicts).
gz-launch/main is in the same state.

On top of those build errors, they're also looking for the wrong project versions (e.g they look for gz-rendering7 instead of gz-rendering8).

Basically building `Harmonic` via [collection-harmonic.yaml](https://github.com/gazebo-tooling/gazebodistro/blob/master/collection-harmonic.yaml) doesn't yet work out of the box.

Should I wait until Harmonic is buildable?

**Comment by azeey** [comment]
> Should I wait until Harmonic is buildable?
Yes, there are a couple of PRs in flight bumping the major versions of `gz-rendering` and its dependees. These are tracked in https://github.com/gazebo-tooling/release-tools/issues/853. Let's wait till all of those are merged.

**Comment by iche033** [comment]
after https://github.com/gazebosim/gz-gui/pull/512 gets in, everything up to gz-gui should be buildable

**Comment by darksylinc** [comment]
Another rebased and retargeted.

**Comment by darksylinc** [comment]
I need to rebase this PR now that the previous one has been merged.

**Comment by darksylinc** [comment]
Are the tests flaky? Jammy always fails on these two:

>   	 51 - INTEGRATION_minimal_scene (Timeout)
>  	 53 - INTEGRATION_transport_scene_manager (SEGFAULT)

And Focal always succeeded except this last time (I made no significant changes that would cause this, possibly flaky):

>  	 45 - INTEGRATION_camera_tracking (SEGFAULT)

As for Windows CI, I fixed the problems caused by my PR; the new warnings don't look like caused by me (?)

And macOS was failing build, I fixed that. Now it fails various tests. I don't know if those are my fault.

File: src/plugins/minimal_scene/MinimalScene.cc [hunk 4]
```diff
@@ -561,6 +578,87 @@ void GzRenderer::BroadcastKeyPress()
...
+      }
+    }
+  }
+
+  for (const auto &ext : deviceExtensions)
+  {
+    externalDevice.deviceExtensions.push_back(VkExtensionProperties{});
+    VkExtensionProperties &extProp = externalDevice.deviceExtensions.back();
+    strncpy(extProp.extensionName, ext.toStdString().c_str(),
+            VK_MAX_EXTENSION_NAME_SIZE);
```
Comment (iche033) [inline src/plugins/minimal_scene/MinimalScene.cc:656]: there is a compile warning here:

```
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘void fillQtDeviceExtensionsToOgre(gz::rendering::v8::GzVulkanExternalDevice&)’ at /home/ian/code/gz_h_ws/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:655:12,
    inlined from ‘std::string gz::gui::plugins::GzRenderer::Initialize(gz::gui::plugins::RenderThreadRhi&)’ at /home/ian/code/gz_h_ws/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:711:35:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:34: warning: ‘char* __builtin_strncpy(char*, const char*, long unsigned int)’ specified bound 256 equals destination size [-Wstringop-truncation]
   95 |   return __builtin___strncpy_chk (__dest, __src, __len,
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
   96 |                                   __glibc_objsize (__dest));

```sh

**Review by iche033 — COMMENTED** [review]
tested this with https://github.com/gazebosim/gz-sim/pull/1735 on Jammy with Qt 5.13.3. It's working well for me with worlds like `sensors_demo.sdf` and `segmetnation_camera.sdf`. 

One thing I noticed is that gui crashes when I tried to resize the window. If it's not trivial to fix, we can ticket an issue for this.

**Comment by iche033** [comment]
> 51 - INTEGRATION_minimal_scene (Timeout)
> 53 - INTEGRATION_transport_scene_manager (SEGFAULT)
> 45 - INTEGRATION_camera_tracking (SEGFAULT)

I've seen these tests fail from time to time. I think they are flaky. 

> And macOS was failing build, I fixed that. Now it fails various tests. I don't know if those are my fault.

I haven't seen those tests failed before. Not sure why. I restarted a new homebrew build and now it only has one expected test failure: https://build.osrfoundation.org/job/ignition_gui-ci-pr_any-homebrew-amd64/1893/. Will need to keep an eye on this.

File: src/plugins/minimal_scene/MinimalScene.cc (hunk 4 above)
Comment (darksylinc) [reply src/plugins/minimal_scene/MinimalScene.cc:656]: That should be impossible? The macros:

```cpp
#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 2) && QT_CONFIG(vulkan)
#endif
```

Which means it should be excluded from complication with Qt 5.13.3

Anyway, the macro says we are doing this:

```cpp
char data[256];
strncpy( data, somePtr, 256);
```

Which is dangerous because if strlen( somePtr ) >= 256; then data[255] will not contain the null terminator `\0`.

Which is why I do on the next line:

```cpp
extProp.extensionName[VK_MAX_EXTENSION_NAME_SIZE - 1u] = 0;
```

in order to enforce the string is always valid.

Anyway, there may be other ways to fix this warning:

```cpp
strncpy(extProp.extensionName, ext.constData(),
              VK_MAX_EXTENSION_NAME_SIZE - 1u);
```

Comment (iche033) [reply src/plugins/minimal_scene/MinimalScene.cc]: ah typo, I was testing with Qt 5.1**5**.3 not Qt 5.1**3**.3. I added the suggested fix to that line (and left other similar lines untouched) to get rid of the warning.  99f887f

**Review by iche033 — APPROVED** [review]
CI looks good. Tested together with https://github.com/gazebosim/gz-sim/pull/1735:

```
gz sim -v 4 --render-engine-gui-api-backend vulkan  sensors_demo.sdf
```
GUI works for me in both vulkan and OpenGL.
